biahub/track.py

```diff
@@ -0,0 +1,656 @@
+        data_dict[input_arrays_name] = array_apply(input_array, func=function, **kwargs)
+    # Generate foreground mask
+    click.echo("Generating foreground mask...")
+    foreground_func_details = tracking_functions["foreground"]
+    foreground_function = resolve_function(
+        foreground_func_details.function
+    )  # Uses function mapping
+
+    fg_input_arrays = [data_dict[name] for name in foreground_func_details.input_arrays]
+    fg_kwargs = foreground_func_details.kwargs if foreground_func_details.kwargs else {}
+
+    foreground_mask = array_apply(*fg_input_arrays, func=foreground_function, **fg_kwargs)
+
+    # Generate contour gradient map
+    click.echo("Generating contour gradient map...")
+    contour_func_details = tracking_functions["contour"]
+    contour_function = resolve_function(contour_func_details.function)  # Uses function mapping
+
+    contour_input_arrays = [data_dict[name] for name in contour_func_details.input_arrays]
+    contour_kwargs = contour_func_details.kwargs if contour_func_details.kwargs else {}
+
+    contour_gradient_map = array_apply(
+        *contour_input_arrays, func=contour_function, **contour_kwargs
+    )
+
+    return foreground_mask, contour_gradient_map
+
+
+def ultrack(
+    tracking_config: MainConfig,
+    foreground_mask: ArrayLike,
+    contour_gradient_map: ArrayLike,
+    scale: Union[Tuple[float, float], Tuple[float, float, float]],
+    databaset_path,
+):
+    """
+    Perform object tracking using the ultrack library.
+
+    This function tracks objects based on a provided foreground mask and contour gradient map
+    using the specified tracking configuration. The results include labeled tracking data,
+    a DataFrame containing track information, and a graph representation of the tracks.
+
+    Parameters:
+    - tracking_config (MainConfig): Configuration settings for the tracking process.
+    - foreground_mask (ArrayLike): Binary or probability mask indicating detected objects.
+    - contour_gradient_map (ArrayLike): Gradient-based contour map used for tracking refinement.
+    - scale (Union[Tuple[float, float], Tuple[float, float, float]]): Scale factors for spatial
+      resolution, given as (Y, X) or (Z, Y, X) depending on the dataset.
+    - databaset_path (Path): Directory where tracking results and configurations will be saved.
+
+    Returns:
+    - np.ndarray: Labeled tracking results in an OME-Zarr format.
+    - pd.DataFrame: DataFrame containing tracking information, including object IDs, positions,
+      and frame associations.
+    - networkx.Graph: Graph representation of object tracks, useful for lineage and connectivity
+      analysis.
+
+    Notes:
+    - The function modifies `tracking_config` to set the working directory for results storage.
+    - The function also saves the tracking configuration in a TOML file for reproducibility.
+    """
+
+    cfg = tracking_config
+
+    cfg.data_config.working_dir = databaset_path
```

Comment:
Using a shared storage as the working directory with the default `SQLite` backend can be a bit problematic.
The shared storage is often not fast enough to keep up with the multiple workers, and the database might get locked.
I suggest using something like `/tmp/{hash}` (or even `/dev/shm/{hash}`) and move the `data.db` and `metadata.toml`, and clean it afterwards.